neural_lam/datastore/mllam.py

```diff
@@ -0,0 +1,355 @@
+# Standard library
+import shutil
+import warnings
+from pathlib import Path
+from typing import List
+
+# Third-party
+import cartopy.crs as ccrs
+import mllam_data_prep as mdp
+import xarray as xr
+from loguru import logger
+from numpy import ndarray
+
+# Local
+from .base import BaseCartesianDatastore, CartesianGridShape
+
+
+class MLLAMDatastore(BaseCartesianDatastore):
+    """Datastore class for the MLLAM dataset."""
+
+    def __init__(self, config_path, n_boundary_points=30, reuse_existing=True):
+        """
+        Construct a new MLLAMDatastore from the configuration file at
+        `config_path`. A boundary mask is created with `n_boundary_points`
+        boundary points. If `reuse_existing` is True, the dataset is loaded
+        from a zarr file if it exists (unless the config has been modified
+        since the zarr was created), otherwise it is created from the
+        configuration file.
+
+        Parameters
+        ----------
+        config_path : str
+            The path to the configuration file, this will be fed to the
+            `mllam_data_prep.Config.from_yaml_file` method to then call
+            `mllam_data_prep.create_dataset` to create the dataset.
+        n_boundary_points : int
+            The number of boundary points to use in the boundary mask.
+        reuse_existing : bool
+            Whether to reuse an existing dataset zarr file if it exists and its
+            creation date is newer than the configuration file.
+
+        """
+        self._config_path = Path(config_path)
+        self._root_path = self._config_path.parent
+        self._config = mdp.Config.from_yaml_file(self._config_path)
+        fp_ds = self._root_path / self._config_path.name.replace(
+            ".yaml", ".zarr"
+        )
+
+        self._ds = None
+        if reuse_existing and fp_ds.exists():
+            # check that the zarr directory is newer than the config file
+            if fp_ds.stat().st_mtime > self._config_path.stat().st_mtime:
+                self._ds = xr.open_zarr(fp_ds, consolidated=True)
```

Comment:
Basing this on modification time is simple, but could lead to all kinds of annoying situations. In particular, if you go in and edit the config, then change your mind and edit it back you still need to rebuild your whole dataset.

Can we not base this comparison on the content of the config rather than the file properties? Since `mdp.Config` is a dataclass it should already have a useable comparison operator I think?

Reply:
Yes, that is a good point. That would require the addition of a feature to `mllam-data-prep` that stores the config that the dataset was created from within the dataset itself (for example serialised into yaml and saved as an attribute). Until that is added how about I change it so we simply issue a warning for now if the config file modification timestamp is more recent that the dataset itself? That would be instead of deleting and recreating the dataset we simply issue a warning.

Reply:
Yes, I think that is fine for now. 

And I think serializing the yaml in data-prep would be a good solution for this later.

Reply:
I will create a separate issue for this smart handling of existing zarr archives. For now, the old zarr archive will be used and a warning will be displayed. The user must manually move the old archive first.